### PR TITLE
add PCL pricing

### PR DESCRIPTION
**Pull request by bobthebuidlr**

### CODE REVIEW

File: contracts/oracle/base/src/helpers.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        .checked_mul(Decimal256::from(price).sqrt())?;
+    let lp_price_model = Decimal::try_from(lp_price_model_256)?;
+
+    // Need to use Uint256 because coin0_amount * price + coin1_amount may overflow the 128-bit limit
+    // E.g. 1000 BTC + 21000 ETH in a pool, with a price of 65000 and 3000:
+    // price = 650 / 0.0000000000003 = 1_267_000_000_000_000
+    // 1_000_000_000_000 * 1_267_000_000_000_000 + 21_000_000_000_000_000_000_000 > Uint128::MAX
+    let tvl_real =
+        Uint256::from(coin0_amount) * Decimal256::from(price) + Uint256::from(coin1_amount);
+    let lp_price_real_256 = Decimal256::from_ratio(tvl_real, total_shares);
```
Comment (piobab) [inline contracts/oracle/base/src/helpers.rs:43]: Maybe better `checked_from_ratio`

File: contracts/oracle/base/src/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    let lp_price_real = Decimal::try_from(lp_price_real_256)?;
+
+    Ok(min(lp_price_model, lp_price_real))
```
Comment (piobab) [inline contracts/oracle/base/src/helpers.rs:46]: I think we are missing conversion back to `uusd` denomination. LP is priced in terms of coin1 so we should multiply the `min` result by coin1 pyth price

File: contracts/oracle/wasm/tests/tests/test_price_source.rs [hunk 3]
```diff
@@ -997,3 +997,91 @@ pub fn test_validate_and_query_astroport_xyk_lp_price_source(
         .assert_price_source(&lp_denom, price_source)
         .assert_price_almost_equal(&lp_denom, expected_price, Decimal::percent(1));
 }
+
+#[test_case(PairType::Custom("concentrated".to_string()), &["uatom","untrn"], Some(Decimal::from_str("8.86506356").unwrap()), Some(Decimal::from_str("0.97696221").unwrap()), [1171210862745u128, 12117922358503u128], &[6,6]; "PCL, 6:6 decimals")]
+#[test_case(PairType::Custom("concentrated".to_string()), &["uatom","untrn"], Some(Decimal::from_str("821123123435412349.73564").unwrap()), Some(Decimal::from_str("0.97696221").unwrap()), [923752936745723845u128, 12117922358503u128], &[6,6]; "PCL, [6, 6] decimals Uint128 overflow)")]
+#[test_case(PairType::Custom("concentrated".to_string()), &["utia","untrn"], Some(Decimal::from_str("0.000000000000000585").unwrap()), Some(Decimal::from_str("0.97696221").unwrap()), [92347562936745723845u128, 12117922358503u128], &[18,6]; "PCL, [18, 6] decimals")]
```
Comment (piobab) [inline contracts/oracle/wasm/tests/tests/test_price_source.rs:1003]: Could you add test case simulating low price and low token amount so we check if there is no rounding issue for small numbers?

File: contracts/oracle/base/src/helpers.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+use std::{cmp::min, str::FromStr};
```
[the remark this answers is not in the document]
Comment (bobthebuidlr) [reply contracts/oracle/base/src/helpers.rs]: Moved the pricing to wasm directory

File: contracts/oracle/base/src/helpers.rs (hunk 2 above)
Comment (bobthebuidlr) [reply contracts/oracle/base/src/helpers.rs]: You're right. I added the conversion now

File: contracts/oracle/wasm/src/helpers.rs [hunk 5]
```diff
@@ -255,3 +274,51 @@ pub fn validate_astroport_xyk_lp_pool(
...
+    // price = 650 / 0.0000000000003 = 1_267_000_000_000_000
+    // 1_000_000_000_000 * 1_267_000_000_000_000 + 21_000_000_000_000_000_000_000 > Uint128::MAX
+    let tvl_real =
+        Uint256::from(coin0_amount) * Decimal256::from(price) + Uint256::from(coin1_amount);
+    let lp_price_real_256 = Decimal256::checked_from_ratio(tvl_real, total_shares)?;
+    let lp_price_real = Decimal::try_from(lp_price_real_256)?;
+
+    let pcl_lp_price = min(lp_price_model, lp_price_real);
+
+    // Oracle prices need ot be denominated in usd. As the PCL calculations happen with coin1
```
Comment (piobab) [inline contracts/oracle/wasm/src/helpers.rs:319]:
```suggestion
    // Oracle prices need to be denominated in uusd. As the PCL calculations happen with coin1
```

File: contracts/oracle/wasm/src/helpers.rs [hunk 6]
```diff
@@ -255,3 +274,51 @@ pub fn validate_astroport_xyk_lp_pool(
...
+    // as the denominator, this needs to be converted to USD.
```
Comment (piobab) [inline contracts/oracle/wasm/src/helpers.rs:320]:
```suggestion
    // as the denominator, this needs to be converted to uusd.
```

**Review by piobab — APPROVED** [review]
LGTM.

Before merge, please wait for @markonmars review.

**Review by markonmars — APPROVED** [review]

File: contracts/oracle/wasm/src/helpers.rs [hunk 7]
```diff
@@ -255,3 +274,94 @@ pub fn validate_astroport_xyk_lp_pool(
...
+pub fn compute_pcl_lp_price_real(
+    coin0_decimals: u8,
+    coin1_decimals: u8,
+    coin0_amount: Uint128,
+    coin1_amount: Uint128,
+    coin0_price: &Decimal,
+    coin1_price: &Decimal,
+    total_shares: Uint128,
+) -> ContractResult<Decimal> {
+    // As the oracle returns prices for whole units, we need to adjust the amount to the correct
```
Comment (piobab) [inline contracts/oracle/wasm/src/helpers.rs:349]: "whole units" - do you mean price for NTRN (not untrn)?

Comment (bobthebuidlr) [reply contracts/oracle/wasm/src/helpers.rs:349]: This was incorrect indeed. I have adjusted to accompany that the oracle returns uusd values.

**Review by piobab — APPROVED** [review]

**Comment by piobab** [comment]
@bobthebuidlr LGTM! Would be nice to put somewhere Risk spreadsheet with numbers. Maybe in `tests/files` directory in `wasm` contract.

File: contracts/oracle/wasm/src/helpers.rs [hunk 8]
```diff
@@ -255,3 +274,94 @@ pub fn validate_astroport_xyk_lp_pool(
 
     Ok(())
 }
+
+pub fn compute_pcl_lp_price(
```
Comment (piobab) [inline contracts/oracle/wasm/src/helpers.rs:278]: I think we should move `compute_pcl_lp_price` and `compute_pcl_lp_price_model` to `lp_pricing` file. IMO it is better place.

File: contracts/oracle/wasm/src/helpers.rs [hunk 9]
```diff
@@ -255,3 +274,94 @@ pub fn validate_astroport_xyk_lp_pool(
...
+    coin1_amount: Uint128,
+    coin0_price: &Decimal,
+    coin1_price: &Decimal,
+    total_shares: Uint128,
+) -> ContractResult<Decimal> {
+    // Need to use Decimal256 because coin0_amount * price + coin1_amount may overflow the Decimal limit
+    // E.g. 1000 BTC + 21000 ETH in a pool, with a price of 65000 and 3000:
+    // price = 650 / 0.0000000000003 = 1_267_000_000_000_000
+    // 1_000_000_000_000 * 1_267_000_000_000_000 + 21_000_000_000_000_000_000_000 > Decimal::MAX
+    let tvl_real = Decimal256::from_ratio(coin0_amount, 1u128) * Decimal256::from(*coin0_price)
```
Comment (piobab) [inline contracts/oracle/wasm/src/helpers.rs:360]: We could simplify:
```
let tvl_real = coin0_amount.checked_mul_floor(*coin0_price)? + coin1_amount.checked_mul_floor(*coin1_price)?;
    let lp_price_real = Decimal::checked_from_ratio(tvl_real, total_shares)?;
```

File: contracts/oracle/wasm/src/helpers.rs [hunk 10]
```diff
@@ -255,3 +274,94 @@ pub fn validate_astroport_xyk_lp_pool(
...
+        .checked_mul(coin0_price_scaled.checked_mul(coin1_price_scaled)?.sqrt())?;
+    let lp_price_model = Decimal::try_from(lp_price_model_256)?;
+
+    Ok(lp_price_model)
+}
+
+pub fn compute_pcl_lp_price_real(
+    coin0_amount: Uint128,
+    coin1_amount: Uint128,
+    coin0_price: &Decimal,
```
Comment (piobab) [inline contracts/oracle/wasm/src/helpers.rs:352]: We can remove references for Decimals

File: contracts/oracle/wasm/src/helpers.rs (hunk 9 above)
Comment (bobthebuidlr) [reply contracts/oracle/wasm/src/helpers.rs:360]: Oh wow yes that makes sense. Much simpler too 😅

File: contracts/oracle/wasm/src/helpers.rs (hunk 8 above)
Comment (bobthebuidlr) [reply contracts/oracle/wasm/src/helpers.rs:278]: Well I think technically the compute functions could belong in both. But I'm fine moving the pcl lp calcs to the `lp_pricing`

**Pull request merged**